### PR TITLE
[MP Rewrite Part 1] Break out and clean up BallState

### PR DESCRIPTION
**Pull request by kylestach**
This substantially cleans up the ball state code, and moves it from
`SystemState` (which we are moving away from in general) to `WorldState`
(which is restricted to physics-only information.

This is not expected to pass CI.

### CODE REVIEW

File: soccer/WorldState.hpp [hunk 1]
```diff
@@ -14,12 +20,80 @@ struct RobotState {
...
+     *  if it is not nullptr.
+     * @return the duration until the ball stops.
+     */
+    [[nodiscard]] RJ::Seconds query_stop_time(
+        Geometry2d::Point* out = nullptr) const;
+
+    /**
+     * Query the time before the ball goes a certain distance.
+     */
+    [[nodiscard]] std::optional<RJ::Seconds> query_seconds_to_dist(
```
Comment (JNeiger) [inline soccer/WorldState.hpp:90]: Needs a @param

Comment (JNeiger) [reply soccer/WorldState.hpp:90]: Why is this optional?

**Comment by kylestach** [comment]
Closing because I actually need to make this branch exist on this fork.

File: soccer/gameplay/robocup-py.cpp [hunk 2]
```diff
@@ -144,19 +170,18 @@ void OurRobot_move_to_end_vel(OurRobot* self, Geometry2d::Point* endPos,
...
-    if (to == nullptr) throw NullArgumentException("to");
+    if (to == nullptr) {
+        throw NullArgumentException("to");
+    }
     self->move(*to);
 }
 
-void OurRobot_settle(OurRobot* self) { self->settle(std::nullopt); }
-
-void OurRobot_settle_w_bounce(OurRobot* self, Geometry2d::Point* bounceTarget) {
```
Comment (JNeiger) [inline soccer/gameplay/robocup-py.cpp:153]: This breaks one touches

File: soccer/planning/DynamicObstacle.hpp [hunk 3]
```diff
@@ -1,54 +1,13 @@
...
-    const Path* getPath() const { return path; }
-
-    // Radius = radius of obstacle
-    float getRadius() const { return radius; }
-
-    std::shared_ptr<Geometry2d::Circle> getStaticObstacle() const {
-        return staticObstacle;
-    }
+class Trajectory;
+struct DynamicObstacle {
```
Comment (JNeiger) [inline soccer/planning/DynamicObstacle.hpp:7]: Could you add comments here describing what this is and how to use it

File: soccer/planning/DynamicObstacle.hpp [hunk 4]
```diff
@@ -1,54 +1,13 @@
 #pragma once
-
-#include <Geometry2d/Point.hpp>
 #include <Geometry2d/Circle.hpp>
-#include "planning/MotionInstant.hpp"
-#include "Utils.hpp"
-#include "Geometry2d/Shape.hpp"
+#include <memory>
```
Comment (JNeiger) [inline soccer/planning/DynamicObstacle.hpp:3]: This one can be removed

**Review by JNeiger — COMMENTED** [review]
There's no source for the ball state and some other things. Not sure if that was intended or not

**Comment by JNeiger** [comment]
Lmao too quick

File: soccer/WorldState.hpp (hunk 1 above)
Comment (kylestach) [reply soccer/WorldState.hpp:90]: It may never actually travel that distance.

**Pull request closed without merging**